sw/device/tests/sram_ctrl_subword_access_test.c

```diff
@@ -0,0 +1,175 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_sram_ctrl.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+#define SRAM_CTRL_TEST_DATA_SIZE_WORDS 3
+#define SRAM_CTRL_TEST_DATA_SIZE_BYTES (SRAM_CTRL_TEST_DATA_SIZE_WORDS * 4)
+
+OTTF_DEFINE_TEST_CONFIG();
+
+static const uint32_t kStatusRegMask = kDifSramCtrlStatusBusIntegErr |
+                                       kDifSramCtrlStatusInitErr |
+                                       kDifSramCtrlStatusEscalated;
+
+static const uint32_t kRandomData[SRAM_CTRL_TEST_DATA_SIZE_WORDS] = {
+    0x6b4abfae, 0x63bdb6e7, 0x87f99b1a};
+
+// Buffer to allow the compiler to allocate a safe area in Main SRAM where
+// we can do the write/read test without the risk of clobbering data
+// used by the program.
+OT_SECTION(".data")
+static volatile uint32_t sram_main_buffer[SRAM_CTRL_TEST_DATA_SIZE_WORDS];
+
+/// Write `kRandomData` to SRAM, then read back a u8, u16, u32, and u64 from
+/// each possible byte offset.
+static void read_subwords_check(mmio_region_t region) {
+  mmio_region_memcpy_to_mmio32(region, 0, kRandomData,
+                               SRAM_CTRL_TEST_DATA_SIZE_BYTES);
+
+  // Check byte reads.
+  for (uint32_t i = 0; i < SRAM_CTRL_TEST_DATA_SIZE_BYTES; ++i) {
+    uint8_t expected = ((uint8_t *)kRandomData)[i];
+    uint8_t got = ((uint8_t *)region.base)[i];
```

Comment:
The array `sram_main_buffer` is declared as `volatile uint32_t` but with the casts throughout this function you're casting away the `volatile`. I imagine that the `mmio` read/write function calls will act as a sufficient barrier to prevent the rest of the code from being optimized away but it might be best to cast to a `volatile` type to be sure.

Reply:
Ah, thank you, I've made all the `sram_main_buffer` pointers that get dereferenced `volatile`.